server/workflowtemplate/workflow_template_server.go

```diff
@@ -69,15 +71,74 @@ func (wts *WorkflowTemplateServer) getTemplateAndValidate(ctx context.Context, n
 
 func (wts *WorkflowTemplateServer) ListWorkflowTemplates(ctx context.Context, req *workflowtemplatepkg.WorkflowTemplateListRequest) (*v1alpha1.WorkflowTemplateList, error) {
 	wfClient := auth.GetWfClient(ctx)
-	options := &v1.ListOptions{}
+	k8sOptions := &v1.ListOptions{}
+
 	if req.ListOptions != nil {
-		options = req.ListOptions
+		k8sOptions = req.ListOptions
 	}
-	wts.instanceIDService.With(options)
-	wfList, err := wfClient.ArgoprojV1alpha1().WorkflowTemplates(req.Namespace).List(ctx, *options)
+
+	resourceVersion := k8sOptions.Continue
+	limit := k8sOptions.Limit
+
+	// kubernetes api will search for all result.
+	// Search whole with limit 0 and save the original limit for custom filtering.
+	k8sOptions.Continue = ""
+	k8sOptions.Limit = 0
+
+	wts.instanceIDService.With(k8sOptions)
+	wfList, err := wfClient.ArgoprojV1alpha1().WorkflowTemplates(req.Namespace).List(ctx, *k8sOptions)
 	if err != nil {
 		return nil, sutils.ToStatusError(err, codes.Internal)
 	}
+
+	// Do name pattern filtering if exist
+	var items []v1alpha1.WorkflowTemplate
+	if req.NamePattern != "" {
+		for _, item := range wfList.Items {
+			if strings.Contains(item.ObjectMeta.Name, req.NamePattern) {
+				items = append(items, item)
+			}
+		}
+	} else {
+		items = wfList.Items
+	}
+
+	// Sort by resourceVersion desc
+	sort.Slice(items, func(i, j int) bool {
+		itemIRV, _ := strconv.Atoi(items[i].ResourceVersion)
+		itemJRV, _ := strconv.Atoi(items[j].ResourceVersion)
+		return itemIRV > itemJRV
+	})
+
+	// Do resourceVersion filtering if continue exist
+	if resourceVersion != "" {
+		newItems := []v1alpha1.WorkflowTemplate{}
+		for _, item := range items {
+			targetRV, _ := strconv.Atoi(item.ResourceVersion)
+			receivedRV, _ := strconv.Atoi(resourceVersion)
+			if targetRV < receivedRV {
+				newItems = append(newItems, item)
+			}
+			items = newItems
+		}
+	}
+
+	// Indexing list by limit count
+	if limit != 0 {
+		endIndex := int(limit)
+		if endIndex > len(items) || limit == 0 {
+			endIndex = len(items)
+		}
+		wfList.Items = items[0:endIndex]
+	} else {
+		wfList.Items = items
+	}
+
+	// Calculate new offset for next batch
+	if limit != 0 && len(wfList.Items) == int(limit) {
+		wfList.ListMeta.Continue = wfList.Items[len(wfList.Items)-1].ResourceVersion
+	}
```

Comment:
@terrytangyuan 
Of course!
We have implemented namePattern filtering.
Since Kubernetes doesn't provide filtering like namePattern, we implemented logic to retrieve results from Kubernetes, then separately filter the results.

In the previous logic, we were utilizing the "continue" option for Kubernetes' built-in pagination.
However, when customizing Kubernetes' results, features like namePattern disrupt the proper usage of "continue" pagination.

Hence, we implemented our own pagination by utilizing Kubernetes' resourceVersion for cursor pagination.

1. Sort in descending order based on resourceVersion.
2. Filter only smaller values compared to the resourceVersion held from the previous page.
3. Index the filtered results based on the limit.
4. Store the resourceVersion for use in the next page.

The pagination logic is implemented in the sequence described above. Thank you.

Also I added a comment about the logic and committed it. https://github.com/argoproj/argo-workflows/pull/11684/commits/e939bb8e81404635d49e76ba3b079692dfe9b282